runtime/macros/src/entry.rs

```diff
@@ -0,0 +1,53 @@
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::{Error, ItemFn};
+
+use super::parse_args::parse_memory_limit;
+
+pub fn main(args: TokenStream, input: TokenStream) -> Result<TokenStream, Error> {
+    let func: ItemFn = syn::parse2(input)?;
+    let memlimit = parse_memory_limit(args)?;
+
+    let fn_sig = &func.sig;
+    if &fn_sig.ident != "main" {
+        return Err(Error::new_spanned(
+            &fn_sig.ident,
+            "function name must be `main`",
+        ));
+    }
+    if fn_sig.asyncness.is_some() {
+        return Err(Error::new_spanned(
+            fn_sig,
+            "`main` function is not allowed to be `async`",
+        ));
+    }
+    if !fn_sig.generics.params.empty_or_trailing() {
+        return Err(Error::new_spanned(
+            &fn_sig.generics,
+            "`main` function is not allowed to have generic parameters",
+        ));
+    }
+    if func.sig.generics.where_clause.is_some() {
+        return Err(Error::new_spanned(
+            &fn_sig.generics.where_clause,
+            "`main` function is not allowed to have a `where` clause",
+        ));
+    }
+
+    Ok(quote! {
+        const _: fn() = main;
```

Comment:
I'm concerned about this. It's correct to do this now, but I don't have a plan about supporting return types on native.
Rust only allows `Result<(), Error>` for main functions. Either we switch from using `main` as a callable function, or introduce some weird syntax like
```rust
nexus_rt::return! { value }; // would print on native and store to memory on riscv, or smth.
```

Maybe it's not that bad?

Reply:
One way or another, some magic in macro will be required to make it work on native